File: src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionCoalesce.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.adapters.JsonAdapter;
import co.nlighten.jsontransform.functions.common.ArgType;
import co.nlighten.jsontransform.functions.common.FunctionContext;
import co.nlighten.jsontransform.functions.common.TransformerFunction;
import co.nlighten.jsontransform.functions.annotations.Aliases;
import co.nlighten.jsontransform.functions.annotations.Documentation;
import co.nlighten.jsontransform.functions.annotations.InputType;
import co.nlighten.jsontransform.functions.annotations.OutputType;

/*
 * For tests
 * @see TransformerFunctionCoalesceTest
 */
@Aliases({"coalesce", "first"})
@Documentation("Returns the first non-null value")
@InputType(ArgType.Array)
@OutputType(value = ArgType.Any, description = "Same as first non-null value")
public class TransformerFunctionCoalesce<JE, JA extends Iterable<JE>, JO extends JE> extends TransformerFunction<JE, JA, JO> {

    public TransformerFunctionCoalesce(JsonAdapter<JE, JA, JO> adapter) {
        super(adapter);
    }

    @Override
    public Object apply(FunctionContext<JE, JA, JO> context) {
        var streamer = context.getJsonElementStreamer(null);
        if (streamer == null || streamer.knownAsEmpty()) return null;
        return streamer.stream()
                .filter(itm -> !adapter.isNull(itm))
                .findFirst()
                .orElse(null);
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionCoalesce.ts
import TransformerFunction from "./common/TransformerFunction";
import FunctionContext from "./common/FunctionContext";
import { isNullOrUndefined } from "../JsonHelpers";

class TransformerFunctionCoalesce extends TransformerFunction {
  constructor() {
    super({});
  }

  override async apply(context: FunctionContext): Promise<any> {
    const streamer = await context.getJsonElementStreamer(null);
    if (streamer == null || streamer.knownAsEmpty()) return null;
    return streamer
      .stream()
      .filter(itm => !isNullOrUndefined(itm))
      .firstOrNull();
  }
}

export default TransformerFunctionCoalesce;
